### ISSUE TITLE
Add cmake support

### ISSUE THREAD

**Issue by xwl** [flycheck/flycheck#865]
For cmake projects, we can get precise command line compile options(flags, include paths, etc) from compile_commands.json generated by cmake.

My current implementation is here:  https://github.com/xwl/flycheck-cmake
Dependency: 
- https://github.com/xwl/cmake-compile-commands
- projectile

It defines two checkers:
- c/c++-cmake
- c/c++-cmake-remote:  // run cmake remotely by ssh

To achieve this, I have to make some minor changes to flycheck.el, shown in above link.  Namely, to make it possible to set the compiler on the fly, instead of a fixed string value.

Are you interested in adding this to flycheck?

**Comment by cpitclaudel** [comment]
Hi @xwl,

Thanks for your work! Yes, we're always interested to improve flycheck. This touches a tiny bit of core Flycheck infrastructure, so I'll let @lunaryorn weight in, but we're always excited to get cool contributions.

Clément.

**Comment by lunaryorn** [comment]
@xwl I'm not sure.  I do not like that change, all the more as it's incomplete and doesn't catch all places where the executable is used.  

Besides, Flycheck already has means to override the executable of a syntax checker.  What does your patch enable that's currently impossible in Flycheck?

**Comment by xwl** [comment]
Different cmake projects can have different compiler setup.  It is hard to set a fixed compiler for all.  And cmake tells you the exact compiler used via exported compile_commands.json.  So i'd like to detect and set compiler automatically. 

The changes to the core are only: 1. remove check against string executable (so make it allow lisp expression) 2. eval exectuable if it is lisp.  It doesn't affect existing core, except losing check for compiler being only string.

I place the changes to the core here:

      
    diff --git a/flycheck.el b/flycheck.el
    index 17e6c3a..f1da106 100644
    --- a/flycheck.el
    +++ b/flycheck.el
    @@ -4360,9 +4360,9 @@ default `:verify' function of command checkers."
    
         (unless command
           (error "Missing :command in syntax checker %s" symbol))
    -    (unless (stringp (car command))
    -      (error "Command executable for syntax checker %s must be a string: %S"
    -             symbol (car command)))
    +    ;; (unless (stringp (car command))
    +    ;;   (error "Command executable for syntax checker %s must be a string: %S"
    +    ;;          symbol (car command)))
         (dolist (arg (cdr command))
           (unless (flycheck-command-argument-p arg)
             (error "Invalid command argument %S in syntax checker %s" arg symbol)))
    @@ -4426,7 +4426,7 @@ the syntax checker definition, if the variable is nil."
    
     Return the full absolute path to the executable of CHECKER, or
     nil if the executable does not exist."
    -  (funcall flycheck-executable-find (flycheck-checker-executable checker)))
    +  (funcall flycheck-executable-find (eval (flycheck-checker-executable checker))))
    
     (defun flycheck-checker-arguments (checker)
       "Get the command arguments of CHECKER."

**Comment by lunaryorn** [comment]
@xwl I'm sorry but that does not answer my question at all.  I didn't ask *what* you changed—I already saw the patch—I asked *why* you made that change.

I'm not convinced that this change is a good idea.  For one, it's not at all complete—please take a careful look at the *whole* Flycheck source to find out where we use the executable, and where we implicitly rely upon it being a string and not an arbitrary Emacs Lisp impression (for instance, in the syntax checker help buffers).  And then `eval` is a bad idea in general.  Flycheck only uses it at a single place, and I'd like to hear a very good case for allowing eval at another.

Are you aware that Flycheck implicitly declares variables to let you change the executable?  For instance. `c/c++-clang` has an option `flycheck-c/c++-clang-executable` which you can set to change the clang executable Flycheck would invoke.  Can't you use this feature?

And if you can't, what do you miss?  Try to show a concrete and simple example that illustrates your need, preferably one that does not involve cmake.

**Comment by xwl** [comment]
flycheck-cmake is meant to be used in cmake projects. I can't give an example without cmake.. cmake itself is just for setting up the build configurations. The thing is, each cmake project can have its own settings, e.g, which compiler is used.  It is not possible to have a fixed string. Or are you suggesting that we just set it with some default value,  when flycheck actually does the compilation, we detect and set the flycheck-c/c++-cmake-executable there?

As for `eval', i see many checkers are using eval, e.g., in flycheck-substitute-argument, it supports a eval form there:

     (`(eval ,form)
      (let ((result (eval form)))

that gives me an impression that eval is used a lot in flycheck.  For the new eval i introduced, it can be probably replaced a funcall, instead.  But then flycheck-command-argument-p needs to match a lisp function, e.g., `((lambda () ,_) t)

**Comment by lunaryorn** [comment]
@xwl Well, then at least give me a *concrete* example, and assume that I do not know anything about `cmake`.  Your explanations are too abstract for me to see why you need to compute the executable the way your suggested patch does, and why you can just use the options that Flycheck already gives you instead.

I don't know how your `flycheck-c/c++-cmake` syntax checker is defined, but presumably, yes, I suggest that you set `flycheck-c/c++-cmake-executable` based on the cmake configuration of the current project.

**Comment by xwl** [comment]
Yes, flycheck-c/c++-cmake-executable is configured by reading cmake generated configs.

e.g., project1 has this config:

    {
      "command": "/usr/bin/clang++  -I/project1/include foo.cpp",
      "file": "/project1/foo.cpp"
    },
    
project2 has different config:
    
    {
      "command": "/usr/bin/g++  -I/project2/include -DQT_CORE_LIB -g bar.cpp",
      "file": "/project2/bar.cpp"
    },

So different projects have their own settings, compiler, build flags(which is usually very long) 

When switching source file buffers between different cmake project, it would be rather inconvenient to change flycheck-c/c++-clang-executable, flycheck-clang-args, etc.  Probably we would have to change them to buffer local variables.  Also, i hope users don't have to handle this manually, then clang would have to recognize cmake config. (and there are other similar tools like cmake..)

In my flycheck-cmake, i defined compiler and compilation flags as:

    (defvar flycheck-cmake-checker-spec
      '(:command
        ((eval (cmake-compile-commands-compiler))
         (eval (cmake-compile-commands-args))

which means for each file in a cmake project, we will always compute its correct settings.

**Comment by lunaryorn** [comment]
@xwl I'm sorry but I don't understand why it'd be "inconvenient".   Presumably there's already code that extracts the compiler name (e.g. `cmake-compile-command-compiler`); why can't you use this function in a hook to set the executable options for the current buffer?

**Comment by xwl** [comment]
(add-hook 'c++-mode-hook '(lambda () (setq flycheck-c/c++-clang-executable (cmake-compile-command-compiler)))

Say we open two files: bufffer foo.cpp use g++, buffer bar.cpp uses clang++. Now, what is the value of flycheck-c/c++-clang-executable? flycheck-c/c++-clang-executable can only have one value( unless it is buffer local variable), and will be set according to last opened file buffer.

**Comment by lunaryorn** [comment]
@xwl Uhm, yes, but why not make the variable buffer local then?  Just use `setq-local` instead of `setq` to achieve this effect.

**Comment by xwl** [comment]
Take clang as example, here is default setting in flycheck:

    (flycheck-define-checker c/c++-clang
      "A C/C++ syntax checker using Clang.
    
    See URL `http://clang.llvm.org/'."
      :command ("clang"
                "-fsyntax-only"
                "-fno-color-diagnostics"    ; Do not include color codes in output
                "-fno-caret-diagnostics"    ; Do not visually indicate the source
                                            ; location
                "-fno-diagnostics-show-option" ; Do not show the corresponding
                                            ; warning group
                "-iquote" (eval (flycheck-c/c++-quoted-include-directory))
                (option "-std=" flycheck-clang-language-standard concat)
                (option-flag "-pedantic" flycheck-clang-pedantic)
                (option-flag "-pedantic-errors" flycheck-clang-pedantic-errors)
                (option "-stdlib=" flycheck-clang-standard-library concat)
                (option-flag "-fms-extensions" flycheck-clang-ms-extensions)
                (option-flag "-fno-exceptions" flycheck-clang-no-exceptions)
                (option-flag "-fno-rtti" flycheck-clang-no-rtti)
                (option-flag "-fblocks" flycheck-clang-blocks)
                (option-list "-include" flycheck-clang-includes)
                (option-list "-W" flycheck-clang-warnings concat)
                (option-list "-D" flycheck-clang-definitions concat)
                (option-list "-I" flycheck-clang-include-path)
                (eval flycheck-clang-args)
                "-x" (eval
                      (pcase major-mode
                        (`c++-mode "c++")
                        (`c-mode "c")))
                ;; Read from standard input
                "-")
      :standard-input t
    
Apart from setting compiler and flycheck-clang-args, users may have to remove all other flags (as it can conflict with cmake project settings, or just duplication), e.g.,  "-fno-color-diagnostics" , etc, and i don't know how to do that for a user.  

All in all, this approach seems requiring both much more changes in flycheck core, and more configurations for the users as well?

With flycheck-cmake, user just needs to config the json files, and in c++-mode-hook, to select correct checker.

**Comment by lunaryorn** [comment]
@xwl I'm sorry, what has clang to do with it?  I thought you were defining your own syntax checker.

I think we're talking past each other, so let me repeat: I was not suggesting that you used the built-in clang syntax checker, but that you keep your own syntax checker, but use it's associated executable option, e.g. `flycheck-c/c++-cmake-executable` to dynamically change the compiler it runs, for instance in a hook like this:

```cl
(add-hook 'c++-mode-hook
                 (lambda () (setq-local flycheck-c/c++-cmake-executable
                            (cmake-compile-command-compiler)))
```

**Comment by xwl** [comment]
Ah, so i misunderstood.  Now i see your point.  It means i just define an "empty" cmake checker, and users have to set both compilers and args.

**Comment by lunaryorn** [comment]
I'm not sure what you mean with "empty", and I don't see why your users should set the compiler and it's arguments.  

I think you can just keep the current definition, and only need to change the way how you handle the executable.

**Comment by xwl** [comment]
OK. Then the problem is when to set this executable? Is flycheck-mode-hook the correct hook to do this? Ideally, this is only needed when selected checker is c/c++-cmake.

**Comment by lunaryorn** [comment]
@xwl Sure, `flycheck-mode-hook` is a good place for this.  Take a look at https://github.com/flycheck/flycheck-haskell to see how we do a similar thing there:  We inspect a Haskell project file and set various options based on its contents.

**Comment by xwl** [comment]
Now i have removed usage of eval. And set it through flycheck-before-syntax-check-hook. 
    https://github.com/xwl/flycheck-cmake
